# 7a Serie de exercicios - FM-293
### Ramon Carlos Gargalhone

In [65]:
import numpy as np
MU = 398600 # Parametro gravitacional

### Exercicio 1

In [66]:
# Input
r0 = 6600 #km
rf = 9900 #km

In [67]:
sigma_f = rf/r0
print(f"Sigma: {sigma_f}")

Sigma: 1.5


Como σ < 11.94, a manobra ótima corresponde a manobra de Hohmann.

Os incrementos de velocidades sao dados por:

In [68]:
# primeiro incremento
dv0 = np.sqrt(MU/r0)*(-1 + np.sqrt(2*sigma_f/(1 + sigma_f)))
print(f"Primeiro incremento: {dv0:.4f} km/s")

#segundo incremento
dvf = np.sqrt(MU/(r0*sigma_f))*(1-np.sqrt(2/(1+sigma_f)))
print(f"Segundo Incremento: {dvf:.4f} km/s")

Cf = dv0 + dvf
print(f"Delta V total: Cf = |∆v0| + |∆vf| = {Cf:.4f} km/s")


Primeiro incremento: 0.7417 km/s
Segundo Incremento: 0.6699 km/s
Delta V total: Cf = |∆v0| + |∆vf| = 1.4116 km/s


O semi-eixo maior e a excentricidade da órbita de transferência

In [69]:
#semi-eixo maior:
a = (r0+rf)/2
print(f"Semi-eixo maior a: {1:.4f} km")

#excentricidade
e = (rf-r0)/(rf+r0)
print(f"Excentricidade: {e:.4f}")



Semi-eixo maior a: 1.0000 km
Excentricidade: 0.2000


duração da manobra é: 

In [70]:
#duracao da manobra
delta_t = (np.pi/2)*np.sqrt((r0+rf)**3/(2*MU))
#tempo em hh:mm:ss
delta_t_h = int(delta_t/3600)
delta_t_m = int((delta_t % 3600)/60)
delta_t_s = (delta_t % 60)
print(f"Tempo de voo: {delta_t_h}hs {delta_t_m} min {delta_t_s:.2f}s")

Tempo de voo: 1hs 2 min 8.74s


A manobra de menor consumo de combustível - mínimo delta-v - que transfere o satélite de uma órbita circular de 6600 para uma órbita circular a 9900 km é realizada mediante a aplicação de dois impulsos: o primeiro impulso, aplicado tangencialmente à órbita circular inicial, fornece um incremento de velocidade de 0.74174 km/s e coloca o satélite no pericentro de uma órbita elíptica de semi-eixo maior 8250.0 km e excentricidade 0.2. O segundo impulso, aplicado tangencialmente à órbita de transferência em seu apocentro, fornece um incremento de velocidade de 0.66989 km/s e coloca o satélite na órbita circular final. O tempo de voo para a realização da manobra é de 1 hs 02 min e 8.74s.

### Exercicio 2

In [71]:
# Input
r0 = 6600 #km
rf = 90000 #km

In [72]:
sigma_f = rf/r0
print(f"Sigma: {sigma_f:.4f}")

Sigma: 13.6364


Como σ > 11.94, a manobra ótima corresponde a bi-parabólica.

O incremento de velocidade no primeiro impulso é dado por

In [73]:
dv0 = np.sqrt(MU/r0)*(-1 + np.sqrt(2))
print(f"Primeiro incremento: {dv0:.4f} km/s")


Primeiro incremento: 3.2190 km/s


O incremento de velocidade no segundo impulso é dado por:

In [74]:
#segundo incremento
dvf = np.sqrt(MU/(rf))*(np.sqrt(2) -1)
print(f"Segundo Incremento: {dvf:.4f} km/s")

Segundo Incremento: 0.8717 km/s


A velocidade característica da manobra:
Cf = |∆v0| + |∆vf| 

In [75]:
Cf = dv0 + dvf
print(f"Delta V total: {Cf:.4f} km/s")

Delta V total: 4.0907 km/s


Semi-latus rectum para bi-parabolica:

In [76]:
p = r0*2
print(f"Semi-latus rectum parabolica 1: {p:.4f} km")

p = rf*2
print(f"Semi-latus rectum parabolica 2: {p:.4f} km")

Semi-latus rectum parabolica 1: 13200.0000 km
Semi-latus rectum parabolica 2: 180000.0000 km


A manobra de mínimo delta-v que transfere o satélite de uma órbita circular de 6600 km de raio para uma órbita circular de 90000 km de raio é: 

Dois impulsos finitos e de um impulso intermediário infinitesimal: o primeiro impulso, aplicado tangencialmente à órbita circular inicial, fornece um incremento de velocidade de 3.2190 km/s e coloca o satélite no pericentro de uma órbita parabólica de semi-latus rectum 13200 km. O impulso intermediário, aplicado a uma distância infinita, coloca o satélite em uma segunda órbita parabólica de semi-latus rectum 180000 km. O segundo impulso finito, aplicado tangencialmente à segunda órbita parabólica em seu pericentro, fornece um incremento de velocidade de 0.87171 km/s, desacelerando o satélite e colocando-o na órbita circular final. 

### Exercicio 3

In [77]:
# Input
r0 = 7000
THETA_DEG = 45
THETA = np.radians(THETA_DEG)

Para 38.94 <= Θ <= 60, temos uma manobra com 3 impulsos. Calculando σ∗

In [78]:
sigma = np.sin(THETA/2)/(1 - 2*np.sin(THETA/2))
print(f"sigma: {sigma:.4f}")

sigma: 1.6310


Calculando a excentricidade da órbita de transferência:

In [79]:
e1 = (sigma - 1)/(sigma + 1)
print(f"Excentricidade manobra intermediaria: {e1:.4f}")

Excentricidade manobra intermediaria: 0.2398


Os incrementos de velocidade iniciais e finais serão iguais, |∆v0| = |∆vf|

In [80]:
dv0 = np.sqrt(MU/r0)*(np.sqrt(1+e1) - 1)
print(f"deltaV manobra inicial = final: {dv0:.4f} km/s")
dvf = dv0

deltaV manobra inicial = final: 0.8563 km/s


Onde o raio da manobra intermediária será:

In [81]:
r = r0*sigma
print(f"raio manobra intermediaria: {r:.4f} km")

raio manobra intermediaria: 11416.9042 km


Manobra intermediaria:

In [82]:
dvi = 2*np.sqrt(MU/r)*np.sqrt(1-e1)*np.sin(THETA/2)
print(f"deltaV manobra inermediaria: {dvi:.4f} km/s")

deltaV manobra inermediaria: 3.9429 km/s


Finalmente, a velocidade característica necessária será:
Cf = |∆v0| + |∆vi| + |∆vf|

In [83]:
Cf = abs(dv0) + abs(dvf) + abs(dvi)
print(f"deltaV Total 3 impulsos {Cf:.4f} km/s")

deltaV Total 3 impulsos 5.6555 km/s


### Exercicio 4

In [84]:
# Input
r0 = 10000
THETA_DEG = 30
THETA = np.radians(THETA_DEG)

Para 0 <= Θ <= 38.94, temos uma manobra com apenas 1 impulso.
O incremento de velocidade necessário para realizar a manobra de mudança de plano orbital com um único impulso:

In [85]:
delta_v = 2*np.sqrt(MU/r0)*np.sin(THETA/2)
print(f"deltaV 1 impulso {delta_v:.4f} km/s")

deltaV 1 impulso 3.2681 km/s


### Exercicio 5

In [86]:
# Input
r0 = 12000
THETA_DEG = 75
THETA = np.radians(THETA_DEG)

Para 60 <= Θ <= 180, temos uma manobra com 3 impulsos bi-parabólica, onde o impulso intermediário e infinitesimal.

O incremento de velocidade inicial será:

In [87]:
# case 3
dv0 = np.sqrt(MU/r0)*(-1 + np.sqrt(2))
print(f"Primeiro incremento: {dv0:.4f} km/s")

Primeiro incremento: 2.3873 km/s


O incremento de velocidade final será: Como r0 = rf

In [88]:
#segundo incremento
dvf = np.sqrt(MU/(r0))*(np.sqrt(2) -1)
print(f"Segundo Incremento: {dvf:.4f} km/s")

Segundo Incremento: 2.3873 km/s


Assim, a velocidade característica necessária será: Cf = |∆v0| + |∆vf| 

In [89]:
Cf = dv0 + dvf
print(f"Delta V total: {Cf:.4f} km/s")

print("Impulso intermediario a uma distancia infinita, de valor infinitezimal")

Delta V total: 4.7745 km/s
Impulso intermediario a uma distancia infinita, de valor infinitezimal


### Exercicio 6

In [90]:
# Input
r0 = 7160  # km
rf = 3 * r0
THETA_DEG = 30
THETA = np.radians(THETA_DEG)

Elementos orbitais:

In [91]:
a = (r0 + rf) / 2
e = (rf - r0) / (rf + r0)
print(f"Semi-eixo maior: {a:.4f} km")
print(f"Excentricidade: {e:.4f}")

Semi-eixo maior: 14320.0000 km
Excentricidade: 0.5000


Velocidades nos apsis da órbita de transferência:

In [92]:
v1_abs = np.sqrt((MU * (1 + e)) / r0)
v2_abs = np.sqrt((MU * (1 - e)) / rf)

print(f"Velocidades nos apsis da órbita de transferência: {v1_abs:.4f}, {v2_abs:.4f} km/s")

Velocidades nos apsis da órbita de transferência: 9.1381, 3.0460 km/s


Velocidades nas órbitas inicial e final:

In [93]:
v0 = np.sqrt(MU / r0)
vf = np.sqrt(MU / rf)
print(f"Velocidades nos apsis da órbita de transferência: {v1_abs:.4f}, {v2_abs:.4f} km/s")
print(f"Velocidades nas órbitas circular inicial e final: {v0:.4f}, {vf:.4f} km/s")

Velocidades nos apsis da órbita de transferência: 9.1381, 3.0460 km/s
Velocidades nas órbitas circular inicial e final: 7.4613, 4.3078 km/s


Aplicando o método de Newton-Raphson: α∗

In [94]:
# Função dDeltav(alpha1)
def f(alpha1):
    term1 = v1_abs * v0 * np.sin(alpha1)
    term1_2 = np.sqrt(v2_abs**2 + vf**2 - 2*v2_abs*vf*np.cos(THETA - alpha1))
    term2 = v2_abs * vf * np.sin(THETA - alpha1)
    term2_2 = np.sqrt(v1_abs**2 + v0**2 - 2*v1_abs*v0*np.cos(alpha1))
    return term1 * term1_2 - term2 * term2_2

# Derivada dDeltav/dalpha1
def df(alpha1, h=1e-6):
    return (f(alpha1 + h) - f(alpha1 - h)) / (2 * h)

# Metodo de Newton-Raphson
def newton_raphson(alpha0, tol=1e-8, max_iter=100):
    alpha = alpha0
    for i in range(max_iter):
        f_val = f(alpha)
        df_val = df(alpha)
        if abs(df_val) < 1e-12:
            print("Derivada próxima de zero, parada.")
            break
        alpha_new = alpha - f_val / df_val
        if abs(alpha_new - alpha) < tol:
            print(f"Convergiu em {i+1} iterações.")
            return alpha_new
        alpha = alpha_new
    raise RuntimeError("Não convergiu.")

alpha0_guess = np.radians(10)  # chute inicial
alpha1_solution = newton_raphson(alpha0_guess)
print(f"Angulo otimo alpha1: {np.degrees(alpha1_solution):.4f} graus")

Convergiu em 5 iterações.
Angulo otimo alpha1: 4.1727 graus


Os incrementos de velocidade em cada um dos impulsos são dados por

In [95]:
delta_v0 = np.sqrt(v1_abs**2 + v0**2 -2*v1_abs*v0*np.cos(alpha1_solution))
print(f"deltaV0: {delta_v0:.4f} km/s")

delta_vf = np.sqrt(v2_abs**2 + vf**2 -2*v2_abs*vf*np.cos(THETA-alpha1_solution))
print(f"deltaVf: {delta_vf:.4f} km/s")

deltaV0: 1.7814 km/s
deltaVf: 2.0526 km/s


A velocidade característica necessária:
Cf = |∆v0| + |∆vf|

In [96]:
print(f"Cf = {(delta_v0+delta_vf):.4f} km/s")

Cf = 3.8340 km/s


### Exercicio 7

Obtenha a expressão para o incremento total de velocidade Deltav
relativo à segunda manobra com três impulsos para o problema de rendez-vous descrita no texto

= a 7.3.2 (hoelker e silber)